SU2_CFD/src/variables/CNEMOEulerVariable.cpp

```diff
@@ -0,0 +1,1466 @@
+    rhoEve += (3.0/2.0) * Ru/Ms[nSpecies-1] * (Tve - Tref[nSpecies-1]);
+  }
+
+  /*--- Set energies ---*/
+  U[nSpecies+nDim]   = rhoE;
+  U[nSpecies+nDim+1] = rhoEve;
+
+  return;
+}
+
+bool CNEMOEulerVariable::GradCons2GradPrimVar(CConfig *config, su2double *U,
+                                              su2double *V, su2double **GradU,
+                                              su2double **GradV) {
+
+  unsigned short iSpecies, iEl, iDim, jDim, iVar, nHeavy, *nElStates;
+  su2double rho, rhoCvtr, rhoCvve, T, Tve, eve, ef, eref, RuSI, Ru;
+  su2double Cvvs, Cves, dCvvs, dCves;
+  su2double thoTve, exptv;
+  su2double An1, Bd1, Bn1, Bn2, Bn3, Bn4, A, B;
+  su2double *rhou, *Grhou2;
+  su2double *xi, *Ms, *Tref, *hf, *thetav;
+  su2double **thetae, **g;
+
+  /*--- Conserved & primitive vector layout ---*/
+  // U:  [rho1, ..., rhoNs, rhou, rhov, rhow, rhoe, rhoeve]^T
+  // V: [rho1, ..., rhoNs, T, Tve, u, v, w, P, rho, h, a, rhoCvtr, rhoCvve]^T
+
+  /*--- Allocate arrays ---*/
+  Grhou2 = new su2double[nDim];
+  rhou   = new su2double[nDim];
+
+  /*--- Determine number of heavy-particle species ---*/
+  if (config->GetIonization()) nHeavy = nSpecies-1;
+  else                         nHeavy = nSpecies;
+
+  /*--- Rename for convenience ---*/
+  rho     = V[RHO_INDEX];
+  rhoCvtr = V[RHOCVTR_INDEX];
+  rhoCvve = V[RHOCVVE_INDEX];
+  T       = V[T_INDEX];
+  Tve     = V[TVE_INDEX];
+  xi      = config->GetRotationModes();
+  Ms      = config->GetMolar_Mass();
+  Tref    = config->GetRefTemperature();
+  hf      = config->GetEnthalpy_Formation();
+  RuSI    = UNIVERSAL_GAS_CONSTANT;
+  Ru      = 1000.0*RuSI;
+  thetav  = config->GetCharVibTemp();
+  g       = config->GetElDegeneracy();
+  thetae  = config->GetCharElTemp();
+  nElStates = config->GetnElStates();
+
+  for (iDim = 0; iDim < nDim; iDim++)
+    for (iVar = 0; iVar < nPrimVarGrad; iVar++)
+      GradV[iVar][iDim] = 0.0;
+
+  for (iDim = 0; iDim < nDim; iDim++) {
+
+    Grhou2[iDim] = 0.0;
+
+    /*--- Species density ---*/
+    for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+      GradV[RHOS_INDEX+iSpecies][iDim] = GradU[iSpecies][iDim];
+      GradV[RHO_INDEX][iDim]          += GradU[iSpecies][iDim];
+    }
+
+    /*--- Velocity ---*/
+    for (jDim = 0; jDim < nDim; jDim++) {
+      GradV[VEL_INDEX+jDim][iDim] = (rho*GradU[nSpecies+jDim][iDim] -
+          rhou[jDim]*GradV[RHO_INDEX][iDim]) / (rho*rho);
+    }
+
+    /*--- Specific Heat (T-R) ---*/
+    GradV[RHOCVTR_INDEX][iDim] = 0.0;
+    for (iSpecies = 0; iSpecies < nSpecies; iSpecies++)
+      GradV[RHOCVTR_INDEX][iDim] += GradU[iSpecies][iDim]*(3.0+xi[iSpecies])/2.0 *Ru/Ms[iSpecies];
+
+    /*--- Temperature ---*/
+    // Calculate the gradient of rho*u^2
+    for (jDim = 0; jDim < nDim; jDim++) {
+      Grhou2[iDim] += 2.0/rho*(rhou[jDim]*GradU[nSpecies+jDim][iDim]) -
+          GradV[RHO_INDEX][iDim]/(rho*rho)*(GradU[nSpecies+jDim][iDim]*
+          GradU[nSpecies+jDim][iDim]);

```

Comment:
Same comment about using pow(x,2), someone like me, who does not know better, needs to go index by index to see that is is actually the thing multiplied by itself, whereas with pow...